docs/man/man3/TTT_Window.3
.TH "TTT::Window" 3 "Thu Mar 9 2023" "Version v1.0.1" "TTT" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TTT::Window \- The 'Window' class initializes the window and everything else (like the default event loop)\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include 'TTT\&.hpp'\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBWindow\fP (int windowWidth, int windowHeight, const sf::String &windowTitle='TTT', uint32_t windowStyle=sf::Style::Default, const sf::ContextSettings &windowContextSettings=sf::ContextSettings())"
.br
.RI "The 'Window' constructor used to create the window using sf::RenderWindow\&. "
.ti -1c
.RI "\fB~Window\fP ()"
.br
.RI "The 'Window' destructor does nothing except for logging that the window is being destroyed\&. "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "sf::RenderWindow \fBwindow\fP"
.br
.RI "The 'window' variable is initialized with the parameters of the 'Window' constructor\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static bool \fBverbose\fP"
.br
.RI "The user can choose whether to show the debug output in the terminal as well as log to a file called 'TTT\&.log' or just log to the file\&. "
.ti -1c
.RI "static bool \fBtoggleFPS\fP"
.br
.RI "The user can choose whether to show the FPS at startup or require the 'F3' key to be pressed to show it\&. "
.ti -1c
.RI "static bool \fBshouldCreateDefaultEventLoop\fP"
.br
.RI "The user can choose whether to create a default immutable event loop or not\&. "
.ti -1c
.RI "static std::string \fBdefaultEventLoopFontFilename\fP = 'resources/fonts/04B_30__\&.TTF'"
.br
.RI "Optional variable for setting default event loop font path\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBdefaultEventLoop\fP ()"
.br
.RI "Creates the default immutable event loop\&. "
.in -1c
.SH "Detailed Description"
.PP 
The 'Window' class initializes the window and everything else (like the default event loop)\&. 

The class is initialized using a 'Window' constructor that takes parameters similar to the parameters of the sf::RenderWindow class constructor\&. 
.PP
Definition at line \fB25\fP of file \fBTTT\&.hpp\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "TTT::Window::Window (int windowWidth, int windowHeight, const sf::String & windowTitle = \fC'TTT'\fP, uint32_t windowStyle = \fCsf::Style::Default\fP, const sf::ContextSettings & windowContextSettings = \fCsf::ContextSettings()\fP)"

.PP
The 'Window' constructor used to create the window using sf::RenderWindow\&. 
.PP
\fBParameters\fP
.RS 4
\fIwindowWidth\fP integer representing the width of the created window\&. 
.br
\fIwindowHeight\fP integer representing the height of the created window\&. 
.br
\fIwindowTitle\fP constant reference to the title of the created window of type sf::String with a default value of 'TTT'\&. 
.br
\fIwindowStyle\fP unsigned 32-bit integer representing the style of the created window with a default value of 'sf::Style::Default'\&. 
.br
\fIwindowContextSettings\fP constant reference to the context settings of the created window of type sf::ContextSettings with a default value of 'sf::ContextSettings()'\&. 
.RE
.PP

.PP
Definition at line \fB38\fP of file \fBTTT_Window\&.cpp\fP\&.
.SS "TTT::Window::~Window ()"

.PP
The 'Window' destructor does nothing except for logging that the window is being destroyed\&. The 'Window' destructor does nothing because it is already done by SFML when using the 'window\&.close()' function\&. 
.PP
Definition at line \fB58\fP of file \fBTTT_Window\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "void TTT::Window::defaultEventLoop ()\fC [private]\fP"

.PP
Creates the default immutable event loop\&. Creates the default immutable event loop that can be turned on or off using a boolean called 'shouldCreateDefaultEventLoop'\&. 
.PP
Definition at line \fB61\fP of file \fBTTT_Window\&.cpp\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "std::string TTT::Window::defaultEventLoopFontFilename = 'resources/fonts/04B_30__\&.TTF'\fC [static]\fP"

.PP
Optional variable for setting default event loop font path\&. If left untouched, the variable will be set to a default value of 'resources/fonts/04B_30__\&.TTF'\&. 
.PP
Definition at line \fB37\fP of file \fBTTT\&.hpp\fP\&.
.SS "bool TTT::Window::shouldCreateDefaultEventLoop\fC [static]\fP"

.PP
The user can choose whether to create a default immutable event loop or not\&. If true, creates a default immutable event loop\&.
.br
If false, does not create a default immutable event loop\&. 
.PP
Definition at line \fB35\fP of file \fBTTT\&.hpp\fP\&.
.SS "bool TTT::Window::toggleFPS\fC [static]\fP"

.PP
The user can choose whether to show the FPS at startup or require the 'F3' key to be pressed to show it\&. If true, the FPS will be shown at startup\&.
.br
If false, to show the FPS, the user will need to press the 'F3' key\&. 
.PP
Definition at line \fB34\fP of file \fBTTT\&.hpp\fP\&.
.SS "bool TTT::Window::verbose\fC [static]\fP"

.PP
The user can choose whether to show the debug output in the terminal as well as log to a file called 'TTT\&.log' or just log to the file\&. If true, displays debug output of the library as well as logs them to a file called 'TTT\&.log'\&.
.br
If false, does not display debug output of the library but logs them to a file called 'TTT\&.log'\&. 
.PP
Definition at line \fB33\fP of file \fBTTT\&.hpp\fP\&.
.SS "sf::RenderWindow TTT::Window::window"

.PP
The 'window' variable is initialized with the parameters of the 'Window' constructor\&. 
.PP
Definition at line \fB31\fP of file \fBTTT\&.hpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for TTT from the source code\&.


CHANGELOG.md
[TOC]

# Release Notes
These are the release notes for each new version of TTT.

## v1.0.1 | Mar. 7, 2023
**Added**
- New static variable `defaultEventLoopFontFilename`

## v1.0.0 | Mar. 5, 2023
- Initial release


src/TTT_Window.cpp
/** \file TTT_Window.cpp
 * \brief Defines the "Window" class.
 */

#include "include/TTT/TTT.hpp"

#include <filesystem>
#include <string>

namespace TTT
{
    /** If true, the FPS will be shown at startup.\n
     * If false, to show the FPS, the user will need to press the "F3" key.
     */
    bool Window::toggleFPS; 

    /** If true, displays debug output of the library as well as logs them to a file called "TTT.log".\n
     * If false, does not display debug output of the library but logs them to a file called "TTT.log".
     */
    bool Window::verbose; 

    /** If true, creates a default immutable event loop.\n
     * If false, does not create a default immutable event loop.
     */
    bool Window::shouldCreateDefaultEventLoop;

    /** If left untouched, the variable will be set to a default value of "resources/fonts/04B_30__.TTF". */
    std::string Window::defaultEventLoopFontFilename = "resources/fonts/04B_30__.TTF";

    /** \brief The "Window" constructor used to create the window using sf::RenderWindow.
     *
     * \param windowWidth integer representing the width of the created window.
     * \param windowHeight integer representing the height of the created window.
     * \param windowTitle constant reference to the title of the created window of type sf::String with a default value of "TTT".
     * \param windowStyle unsigned 32-bit integer representing the style of the created window with a default value of "sf::Style::Default".
     * \param windowContextSettings constant reference to the context settings of the created window of type sf::ContextSettings with a default value of "sf::ContextSettings()".
     */
    Window::Window(int windowWidth, int windowHeight, const sf::String& windowTitle, uint32_t windowStyle, const sf::ContextSettings& windowContextSettings)
        : window(sf::VideoMode(windowWidth, windowHeight), windowTitle, windowStyle, windowContextSettings)
    {
        if (std::filesystem::exists("TTT.log"))
            std::filesystem::remove("TTT.log");

        Log::log(Log::WINDOW, Log::INFO, "Creating window...");

        Log::log(Log::WINDOW, Log::INFO, "Window width: " + std::to_string(windowWidth));
        Log::log(Log::WINDOW, Log::INFO, "Window height: " + std::to_string(windowHeight));
        Log::log(Log::WINDOW, Log::INFO, "Window title: " + windowTitle);

        if (shouldCreateDefaultEventLoop)
            defaultEventLoop();
    }

    /** \brief The "Window" destructor does nothing except for logging that the window is being destroyed.
     *
     * The "Window" destructor does nothing because it is already done by SFML when using the "window.close()" function.
     */
    Window::~Window() { Log::log(Log::WINDOW, Log::INFO, "Destroying window..."); }

    /** Creates the default immutable event loop that can be turned on or off using a boolean called "shouldCreateDefaultEventLoop". */
    void Window::defaultEventLoop()
    {
        Log::log(Log::WINDOW, Log::INFO, "Creating default event loop...");

        window.setFramerateLimit(60);

        sf::Font font = Utils::createFont(defaultEventLoopFontFilename);

        sf::Text text;
        text.setFont(font);
        text.setCharacterSize(26);
        text.setString("If you're seeing this, it means that\nthe default event loop\nhas been initialized and\nis working without any errors.");
        text.setPosition((window.getSize().x / 2) - (text.getLocalBounds().width / 2), (window.getSize().y / 2) - (text.getLocalBounds().height / 2));

        while (window.isOpen())
        {
            sf::Event event;

            while (window.pollEvent(event))
            {
                switch (event.type)
                {
                    case event.Closed:
                        window.close();
                        break;
                    case event.KeyPressed:
                        if (event.key.code == sf::Keyboard::Escape)
                            window.close();
                        break;
                    default:
                        break;
                }
            }

            window.clear();

            window.draw(text);

            /* The "showFPS" function is included so the default event loop has a purpose. Of course it does have a purpose without the "showFPS"
             * function but there's always more it can do. */
            Utils::showFPS(window, font, 24);

            window.display();
        }
    }
}
